contracts/VirtualSwap/Bridge.sol

```diff
@@ -0,0 +1,619 @@
+                mediumSynthKey,
+                expectedMediumSynthAmount,
+                synthOutKey
+            );
+    }
+
+    // Swaps a token from a Saddle's pool to any virtual synth
+    function tokenToSynth(
+        ISwap swap,
+        uint8 tokenFromIndex,
+        bytes32 synthOutKey,
+        uint256 tokenInAmount,
+        uint256 minAmount,
+        address recipient
+    ) external returns (uint256) {
+        // Struct to hold data for this function
+        TokenToSynthInfo memory v =
+            TokenToSynthInfo(IERC20(0), 0, 0, 0, SynthSwapper(0), 0);
+
+        if (recipient == address(0)) {
+            recipient = msg.sender;
+        }
+
+        // Transfer token from msg.sender
+        v.tokenFrom = swap.getToken(tokenFromIndex); // revert when token not found in swap pool
+        v.tokenFrom.safeTransferFrom(msg.sender, address(this), tokenInAmount);
+        tokenInAmount = v.tokenFrom.balanceOf(address(this));
+        v.tokenFrom.approve(address(swap), tokenInAmount);
+
+        // Swaps token to the supported synth in the pool (sETH, sBTC, or sUSD depending on the pool)
+        v.synthIndex = _getSynthIndex(swap); // revert when synth index is not set for given swap address
+        v.mediumSynthAmount = swap.swap(
+            tokenFromIndex,
+            v.synthIndex,
+            tokenInAmount,
+            0,
+            block.timestamp
+        );
+
+        IERC20 mediumSynth = IERC20(_getSynthAddress(swap));
+        v.synthSwapper = new SynthSwapper();
+        mediumSynth.transfer(address(v.synthSwapper), v.mediumSynthAmount);
+
+        // Swap synths via Synthetix network
+        v.destSynthAmount = v.synthSwapper.swapSynth(
+            SYNTHETIX,
+            _getSynthKey(swap),
+            v.mediumSynthAmount,
+            synthOutKey
+        );
+        require(v.destSynthAmount >= minAmount, "Insufficient output");
+
+        // Add the synthswapper to the pending settlement list
+        v.itemId = _addToPendingSynthSettlementList(
+            PendingSynthSettlement(
+                v.synthSwapper,
+                synthOutKey,
+                recipient,
+                false
+            )
+        );
+
+        // Emit TokenToSynth event with relevant data
+        emit TokenToSynth(
+            msg.sender,
+            swap,
+            v.tokenFrom,
+            tokenInAmount,
+            v.destSynthAmount,
+            v.itemId
+        );
+
+        return (v.itemId);
+    }
+
+    function calcSynthToToken(
+        ISwap swap,
+        bytes32 synthInKey,
+        uint8 tokenToIndex,
+        uint256 synthInAmount
+    ) external view returns (uint256) {
+        IExchangeRates exchangeRates =
+            IExchangeRates(SYNTHETIX_RESOLVER.getAddress(EXCHANGE_RATES_NAME));
+
+        uint8 mediumSynthIndex = _getSynthIndex(swap);
+        bytes32 mediumSynthKey = _getSynthKey(swap);
+        require(synthInKey != mediumSynthKey, "use normal swap");
```

Comment:
is this the case when someone tries to do a synth swap of two assets that exist in the same pool?

Reply:
I believe it's the case when you are swapping from a synth in the same pool as the destination asset already (e.g. sBTC -> wBTC).

Reply:
Yep Sunil's comment is right. I can also add support for doing normal swap instead of reverting it. Don't think it's a huge issue though.